keps/sig-network/0031-service-topology.md

```diff
@@ -0,0 +1,259 @@
+
+Kubernetes scheduler can constraining a pod to only be able to run on particular nodes/zones. However, Kubernetes service proxy just randomly picks an available backend for service routing and this one can be very far from the user, so we need a topology-aware service routing solution in Kubernetes. Basically, to find the nearest service backend. In other words, allowing people to configure if ALWAY reach a to local service backend. In this way, they can reduce network latency, improve security, save money and so on. However, because topology is arbitrary, zone, region, rack, generator, whatever, who knows? We should allow arbitrary locality.
+
+`ExternalTrafficPolicy` was added in v1.4, but only for NodePort and external LB traffic. NodeName was added to `EndpointAddress` to allow kube-proxy to filter local endpoints for various future purposes.
+
+Based on our experience of advanced routing setup and recent demo of enabling this feature in Kubernetes, this document would like to introduce a more generic way to support arbitrary service topology.
+
+## Proposal
+
+This proposal builds off of earlier requests to [use local pods only for kube-proxy loadbalancing](https://github.com/kubernetes/kubernetes/issues/7433) and [node-local service proposal](https://github.com/kubernetes/kubernetes/pull/28637). But, this document proposes that not only the particular "node-local" user case should be taken care, but also a more generic way should be figured out.
+
+Locality is an "user-defined" thing. When we set topology key "hostname" for service, we expect node carries different node labels on the key "hostname".
+
+Users can control the level of topology. For example, if someone run logging agent as a daemonset, he can set the "hard" topology requirement for same-host. If "hard" is not met, then just return "service not available". 
+
+And if someone set a "soft" topology requirement for same-host, say he "preferred" same-host endpoints and can accept other hosts when for some reasons local service's backend is not available on some host.
+
+If multiple endpoints satisfy the "hard" or "soft" topology requirement, we will randomly pick one by default. Routing decision is expected to be implemented in L3/4 VIP level such as kube proxy.
+
+
+## Implementation history
+
+### API changes
+
+The user need a way to declare which service is "local service" and what is the "topology key" of "local service".
+
+This will be accomplished through a new type object `ServicePolicy`.
+Endpoint(s) with specify label will be selected by label selector in
+`ServicePolicy`, and `ServicePolicy` will declare the topology policy for those endpoints.
+
+Cluster administrators can configure what services are "local" and what topological they prefer via `ServicePolicy`. `ServicePolicy` is a namespace-scope resource and is strict optional. We can configure policies other than topological reference in `ServicePolicy`, but this proposal will not cover them.
```

Comment:
This is the crux for me - why would cluster admins decide this?   It needs to be the same people deciding the delpolyment (e.g, if I deploy zonal, I want service zonal).

This does raise a point that we really do want Deployment to be more topology aware, so I can have N instances in each zone.  I know there's a KEP for that, don't know the status

Reply:
I remember we have discussed the question on if service policy should be part of service. If we configure it outside of service in a more admin-owned way, it looks more like Istio.

Probably users(not only cluster admins) should be able to configure the service policy? After all, service policy is namespace-scoped in this proposal.

Reply:
yes, I don't think it's an admin thing, it's a service owner thing

Reply:
Agree.  If we had a bunch of policies we wanted to configure or if it was sufficiently complicated, a new resource might be justified, but I don't think it is.